Add using 'Equal' method of structure to test equality

bgaifullin opened the pull request:
If struct implements Equal method it should be used instead of
recursively compare fields, because in some cases recursively
comparing fields returns false result. for example type decimal.Decimal
can have different internal representation for the same values.

bgaifullin commented:
up

bgaifullin commented:
up

bgaifullin commented:
Folks, please review the patch. I cannot integrate package decimal to my project without this PR.

bgaifullin commented:
ping

bgaifullin commented:
Is there any chance that this PR will be merged?

bgaifullin commented:
up

ernesto-jimenez commented on .idea/libraries/GOPATH__testify_.xml:
```diff
@@ -0,0 +1,17 @@
+      <root url="file://$PROJECT_DIR$" />
+    </excluded>
+  </library>
+</component>
```
Remove IDE files

ernesto-jimenez commented on .idea/libraries/Go_SDK.xml:
```diff
@@ -0,0 +1,10 @@
+      <root url="file:///usr/local/go/src" />
+    </SOURCES>
+  </library>
+</component>
```
this one too

ernesto-jimenez commented on assert/equals.go:
```diff
@@ -0,0 +1,159 @@
+		if equal.IsValid() {
+			result := equal.Call([]reflect.Value{v2})
+			return result[0].Bool()
+		}
```
What happens if the `Equal` method has a signature you don't expect?

bgaifullin commented:
@ernesto-jimenez thanks, I have addressed your comments.
Removed ide files.
It will panic if Equal method signature does not fit required, so added check for Equal method signature. Also covered this case by tests.

ernesto-jimenez commented:
@bgaifullin I just realised most of the code is copied and pasted from the standard's library source code.

I'm a bit uncomfortable with forking code from the standard library:
- The license requires keeping the copyright notices, which are missing
- We are copyng all the source but not the tests
- Testify could get out of sync with the stdlib behaviour (e.g: Go 1.8 included a fix on reflect.DeepEqual)

What is the issue aligning your expectation with the result?

Would it be possible to address it without forking `reflect.DeepEqual`?

bgaifullin commented:
@ernesto-jimenez you are right, most of code of function deepEqual is forked from standard library.
My first goal is solving problem with comparing type Decimal (the appropriate PR to shopspring/decimal  is mentioned above), but this PR also actual for any type that has different behaviour between member-wise comparison and comparison via Equal method. for example the default type time. you can have the same time in different Locations and for it the standard DeepEqual returns false, but method time.Equal returns true.
I do not know how to implement this without forking the standard DeepEqual. Comparison vis using method Equal is not related to function DeepEqual. 
Also I cannot implement the separate method for comparison via method Equal, because there can be a complex structures like slices, arrays, map other structures, which contains members which implements Equal method. for the instance: 
```
t1 =: time.Now()
assert.Equal([]time.Time{t1}, []time.Time{t1.In(time.UTC())})
```
This assert should returns true, even local time zone is not UTC 
I also do not like forking functional, but I do not see other ways to implement this.

Any suggestions are welcome.

ernesto-jimenez commented:
@bgaifullin have you considered creating an issue to support this in the standard library?

It seems like it would be the best place to contribute such a change. If they have any objections, I would be interested in learning what those objections are.

bgaifullin commented:
@ernesto-jimenez there is proposal to implement such think in go standard library: https://github.com/golang/go/issues/20444
Unfortunately it was rejected. :(

ernesto-jimenez commented:
@bgaifullin as they point out on that issue, the change would be a breaking change. We would like to avoid that too since we want to keep testify backwards compatible.

Looking at the issue for the stdlib, the objection was to change the behaviour of `DeepEqual`, but they might be willing to consider a proposal to add a new function with that feature.

It might be worth raising a proposal that is backwards compatible.

bgaifullin commented:
@ernesto-jimenez ok, I will try to raise such proposal.

bgaifullin commented:
@ernesto-jimenez you can vote for [proposal ](https://github.com/golang/go/issues/20519)

bgaifullin commented:
@ernesto-jimenez the proposal has been declined. Can you accept this PR, or I should close it ?

bgaifullin commented:
@ernesto-jimenez  What do you think about using third party library to check equality.
for example https://github.com/pierrre/compare or https://github.com/google/go-cmp

jeromefroe commented:
+1 for investigating using `go-cmp` to replace `reflect.DeepEqual`.

This seems particularly relevant given Go 1.9's changes to `time.Time` will likely break many people's tests. Furthermore, I think most people would agree that if the values have an `Equal` method they should be compared using that method before falling back on basic types. 

In regards to backwards compatibility, a major version bump would probably satisfy most people's needs since they can vendor older versions if they so please. In fact, it seems likely to me that there will be more people whose tests are broken because of the changes to `time.Time` in Go 1.9 than people whose tests would be broken by using an equality function which checks for an `Equal` method first.

An alternative option, as well, is to introduce a new equality function for this purpose. Then users can choose when they want to use `Equal` and when they want to use the new function.

ernesto-jimenez commented:
Moving to `go-cmp` would be great, yes.

I've created a specific issue and labeled as `help wanted` https://github.com/stretchr/testify/issues/535

Closing this in favour of that implementation.
